crates/ra_hir_def/src/resolver.rs

```diff
@@ -291,19 +292,26 @@ impl Resolver {
                 }
                 Scope::GenericParams { .. } => continue,
 
-                Scope::ImplDefScope(impl_) if n_segments > 1 => {
+                Scope::ImplDefScope(impl_) => {
                     if first_name == &name![Self] {
-                        let ty = TypeNs::SelfType(*impl_);
-                        return Some(ResolveValueResult::Partial(ty, 1));
+                        if n_segments > 1 {
+                            let ty = TypeNs::SelfType(*impl_);
+                            return Some(ResolveValueResult::Partial(ty, 1));
+                        } else {
+                            return Some(ResolveValueResult::ValueNs(ValueNs::ImplSelf(*impl_)));
+                        }
                     }
                 }
-                Scope::AdtScope(adt) if n_segments > 1 => {
+                Scope::AdtScope(adt) => {
+                    if n_segments == 1 {
+                        // bare `Self` doesn't work in the value namespace in a struct/enum definition
```

Comment:
Is it? What if a struct is a unit struct?

Reply:
If the struct is a unit struct, how will you have code referencing `Self`? (`AdtScope` is in the struct definition...)

But also with a tuple struct, rustc didn't accept it (e.g. `struct A([usize; f(Self([1]))]);`)

Reply:
See e.g. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ec18c3a813abbae4f9522ca69221d865

Reply:
Ah, right, that’s ADT scope...

bors r+